data-model: clarifications for numbers - float & int

rvagg opened the pull request:
Ref: https://github.com/ipld/specs/issues/342

The main thing this does is rule out IEEE 754 specials `NaN`, `Infinity` and `-Infinity` from the data model. 👍 or 👎 to resolve that issue here.

mvdan commented on data-model-layer/data-model.md:
```diff
@@ -85,9 +85,42 @@ Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or
+  of the integer kind to the signed 64-bit range.
+* JavaScript has difficulties safely handling and representing integers outside
+  of the 53-bit unsigned range and differentiating between integers and floats
+	where there is no fractional component.
```
mixing tabs and spaces? :)

rvagg replied:
sigh .. the perils of having JS & Go people share docs, and me being a Go+JS person using the same editor for both, will fix

mvdan commented on data-model-layer/data-model.md:
```diff
@@ -85,9 +85,42 @@ Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or
+Float is a scalar kind and refers to a number with a fractional component,
+represented with a decimal point. This may also include numbers where the
+fractional component is zero (although the distinction between these numbers
+and integers is blurred in some cases—specifically the JavaScript memory model).
```
maybe spaces around the dash? otherwise it almost reads like a single word. but I'm not an expert at technical writing :)

rvagg replied:
in the code font on here it looks like an en-dash but it's an em-dash, which is longer and shouldn't have spaces around it when used in sentences like this. If you look at it in preview it comes out properly: https://github.com/ipld/specs/blob/rvagg/ieee754-specials/data-model-layer/data-model.md

mvdan replied:
gotcha, sounds good. I guess in here it's monospace.

mvdan commented on data-model-layer/data-model.md:
```diff
@@ -85,9 +85,42 @@ Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or
+content-addressable systems in general) due to the broad scope for introducing
+variability in byte representations. Alternative methods for encoding
+fractional numbers with greater precision and less variability should be
+considered where possible.
```
do we want to give some examples here? for example, off the top of my head it's not obvious how I would store a "big int" or a "big float" with arbitrary precision in IPLD. A base-10 string? Some sort of binary encoding in bytes?

rvagg replied:
I've added some examples but would appreciate some sanity checking on them!

mvdan commented on data-model-layer/data-model.md:
```diff
@@ -85,9 +85,42 @@ Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or
+fractional numbers with greater precision and less variability should be
+considered where possible.
+
+There is not a strict 1:1 mapping between the Float kind and any particular
```
But we still do support a subset of IEEE 754 though, right? Otherwise it almost seems like Float is little more than a bunch of arbitrary bytes, which might contain practically anything from a spec point of view.

rvagg replied:
Yes, this is a tricky point - we're talking about the data model here which is about what you interact with programmatically, so (1) that's a step away from codecs, where the rubber hits the road on the bytes->something question, and (2) this is a matter for programming languages and IPLD libraries.

I've added this to try and get closer to the point:

> However, for practical purposes, the IPLD Float kind will be implemented using IEEE 754 primitives and byte layout in most languages and codecs due to its ubiquity.

I hope that helps clarify, along with some other words I've added.

mvdan replied:
Yeah, I think that wording does help.

mvdan reviewed (approved):
SGTM, thanks for working on this. Putting on my "new to IPLD" hat, I have some minor suggestions.

rvagg commented:
Please review my latest commit, I've shuffled things a bit and added more wording to clarify the relationship with IEEE754, added some more to the reasoning behind us not recommending the use of Float and added some examples of possible replacements for it.

mvdan commented on data-model-layer/data-model.md:
```diff
@@ -106,21 +106,43 @@ represented with a decimal point. This may also include numbers where the
+ * The broad scope for introducing variability in byte representations.
+ * The ambiguity introduced in some languages that may cause round-trip
+   discrepancies; specifically JavaScript which does not clearly disambiguate
+	 between "float" and "integer" in its memory model.
```
it happened again :P

mvdan reviewed (approved):
LGTM!

vmx reviewed (approved):
I really like what you've added in the new commit!
